[skip appveyor] Updates for 4.7.0 release (#8266)
kokkos/kokkos: CHANGELOG.md

# CHANGELOG

## 4.7.00

[Full Changelog](https://github.com/kokkos/kokkos/compare/4.6.02...4.7.00)

> 

Reviewer: ```suggestion

### Features:
* Introduce mdspan-based View implementation [\#7427](https://github.com/kokkos/kokkos/pull/7427)
  * This change should be transparent for most users, but it is a complete reimplementation of `View`

```

### Backend and Architecture Enhancements:

#### CUDA:
* Support RDC with Clang 17+ and use new offload driver [\#7831](https://github.com/kokkos/kokkos/pull/7831)
* Add support for AMPERE87 architecture (Jetson Orin Nano) [\#8092](https://github.com/kokkos/kokkos/pull/8092)
* Directly use fp16 mathematical function when available instead of casting back and forth to fp32 [\#7698](https://github.com/kokkos/kokkos/pull/7698)

> * Disable `mdspan`-based view for CUDA 12.9 due to compiler incompatibility[\#8169](https://github.com/kokkos/kokkos/pull/8169)

Reviewer 2: ```suggestion
```

Author: Perhaps this should be moved to the incompatibility section. I'm not sure if this should just be deleted, given that this was a workaround to mitigate a problem for a specific CUDA version.
@cwpearson, any opinion on this?

#### SYCL:
* Add support for Intel DG2 GPUs such as the Arc Alchemist GPUs [\#7922](https://github.com/kokkos/kokkos/pull/7922)

> * Allow requesting out-of-order queues [\#8091](https://github.com/kokkos/kokkos/pull/8091)

Reviewer 2: What is the message here?  This is an IMPL_ option, so technically not something advertised to users to set.

Author: It looks like the message is that `KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES` is internally defined even if `KOKKOS_ENABLE_IMPL_SYCL_OUT_OF_ORDER_QUEUES` was not enabled during configuration.

But yes, those are `IMPL_` options and maybe not needed to be communicated with users.
@masterleinad, are we okay removing this?

Reviewer: Proposal: remove this, but keep something in the release briefing.

* Allow using non-trivially-copyable comparators with oneDPL [\#7538](https://github.com/kokkos/kokkos/pull/7538)
* Directly use fp16 mathematical function when available instead of casting back and forth to fp32 [\#7698](https://github.com/kokkos/kokkos/pull/7698)

#### OpenMPTarget:
* Remove support for non-llvm compilers [\#7973](https://github.com/kokkos/kokkos/pull/7973)

### General Enhancements

> * Introduce mdspan-based View implementation [\#7427](https://github.com/kokkos/kokkos/pull/7427)

Reviewer: ```suggestion
```

Proposing to move this to the top, more as a big awareness issue.

* Add support for SiFive Unmatched Rev B (RISC-V)  [\#7921](https://github.com/kokkos/kokkos/pull/7921) 
* Improve performance of `atomic_op_fetch()` [\#8014](https://github.com/kokkos/kokkos/pull/8014)
* Pass label by reference in all `Kokkos::tools` APIs [\#8065](https://github.com/kokkos/kokkos/pull/8065)
* Compatibility with C+20 modules: Avoid static variables and functions in header files [\#8071](https://github.com/kokkos/kokkos/pull/8071)
* Exit early at initialize with `--kokkos-help` [\#8079](https://github.com/kokkos/kokkos/pull/8079)
* Add Kokkos::SIMD::SVE support for 128-bit and 256-bit SVE [#7807](https://github.com/kokkos/kokkos/pull/7807)
* Add constructors for Random_XorShift*_Pool with execution space argument [\#8039](https://github.com/kokkos/kokkos/pull/8039)
* Implement `nextafter` for fp16 [\#8118](https://github.com/kokkos/kokkos/pull/8118)

> * Enable structured binding return for `parition_space(ExecSpace, Args...)` [\#8114](https://github.com/kokkos/kokkos/pull/8114)

Reviewer 2: ```suggestion
* Change return type of `Experimental::partition_space()` to enable structured binding [\#8114](https://github.com/kokkos/kokkos/pull/8114)```

* Add `constexpr` specifier to `operator==` and `operator!=` for `Kokkos::complex` [\#8254](https://github.com/kokkos/kokkos/pull/8254)

> * Introduce mdspan-based View implementation [\#7427](https://github.com/kokkos/kokkos/pull/7427)

Reviewer 3: I think this should be emphasized more as it can _potentially_ (but not likely) be a breaking change. We may want to put a caveat or something in the changelog and emphasize this more. @crtrott

Author: At minimum, I can move this to the top of the general enhancements section and add it as a key feature in the 4.7 release briefing announcement.
But let me know if we want to emphasize this even more. Then we could perhaps introduce a new section for this for an extra emphasis.

* Remove `nodiscard` attributes from Kokkos simd interface [\#8084](https://github.com/kokkos/kokkos/pull/8084)
* Add support for graph capture (`Cuda`, `HIP` and `SYCL`) [\#7552](https://github.com/kokkos/kokkos/pull/7552)
* Enforce unit launch bound for graph `then` node [\#8192](https://github.com/kokkos/kokkos/pull/8192)
* Add `then_host` to graph [\#8075](https://github.com/kokkos/kokkos/pull/8075)
* Allow building `Kokkos::Experimental::Graph` object directly [\#8108](https://github.com/kokkos/kokkos/pull/8108)

### Build System Changes

> * Set GCC minimum version for cxx 20 [\#8130](https://github.com/kokkos/kokkos/pull/8130)

Reviewer: ```suggestion
* Set GCC minimum version for C++20 to 10.4 [\#8130](https://github.com/kokkos/kokkos/pull/8130)
```

> * Check if compiler/linker can consume the flags set [\#7891](https://github.com/kokkos/kokkos/pull/7891)

Reviewer 2: ```suggestion
* Check at configuration time that compiler/linker can consume the flags set [\#7891](https://github.com/kokkos/kokkos/pull/7891)
```

> * Error out for incompatible RDC and BUILD_SHARED_LIBS [\#8196](https://github.com/kokkos/kokkos/pull/8196)

Reviewer: ```suggestion
* Disallow relocatable device code for shared library build [\#8196](https://github.com/kokkos/kokkos/pull/8196)
```

* Add NVIDIA Blackwell architecture support to the generated makefiles [\#8055](https://github.com/kokkos/kokkos/pull/8055)

> * Add support for building C++20 modules [\#8132](https://github.com/kokkos/kokkos/pull/8132)

Reviewer 2: ```suggestion
```
I think that is just going to confuse people.  I assume you meant to report the configuration option.

Author: I think the idea is to advertise `KOKKOS_ENABLE_EXPERIMENTAL_CXX20_MODULES` option, and that it is applied to Kokkos std algorithms.
Either we can reword this statement to specify that this is only for `Algorithms` for now, or we could wait until all of #8117 is complete and report them as one stack.
@masterleinad, any opinion on this?

> * Set GCC minimum version for cxx 20 [\#8130](https://github.com/kokkos/kokkos/pull/8130)

Reviewer 2: ```suggestion
* Bump GCC minimum required version for C++20 from 10.1 to 10.4 [\#8130](https://github.com/kokkos/kokkos/pull/8130)
```

* Error out for incompatible RDC and BUILD_SHARED_LIBS [\#8196](https://github.com/kokkos/kokkos/pull/8196)

### Incompatibilities (i.e. breaking changes)
* Raise IntelLLVM minimum for SYCL to 2024.2.1 [\#7918](https://github.com/kokkos/kokkos/pull/7918)

> * Disallow `Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF` [\#8021](https://github.com/kokkos/kokkos/pull/8021)

Reviewer 2: Should that one be listed under "Deprecations" instead?

Author: I don't think this is really about deprecation. It does deprecate `Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK` option in that PR, but I'd argue that the focus this PR is about always enforcing the check on dualview `modify_on_[host|device]` flags, which would be a breaking change.

### Deprecations

> * Deprecate KOKKOS_MEMORY_ALIGNMENT, KOKKOS_MEMORY_ALIGNMENT_THRESHOLD [\#7873](https://github.com/kokkos/kokkos/pull/7873)

Reviewer 2: ```suggestion
* Deprecate `KOKKOS_MEMORY_ALIGNMENT[_THRESHOLD]` macros [\#7873](https://github.com/kokkos/kokkos/pull/7873)
```

> * Deprecate Kokkos::MemoryManaged [\#8066](https://github.com/kokkos/kokkos/pull/8066)

Reviewer 2: ```suggestion
* Deprecate `Kokkos::MemoryManaged` alias [\#8066](https://github.com/kokkos/kokkos/pull/8066)
```

Reviewer 2: We need to report that we added a default template argument to the `MemoryTraits` class template.
You can either mention it as the replacement here or as improvement as we did not have a good way to spell it before without spelling out a bitmask value which technically was an implementation detail.

> * Deprecate `KOKKOS_NONTEMPORAL_PREFETCH_[LOAD|STORE]` [\#8146](https://github.com/kokkos/kokkos/pull/8146)

Reviewer 2: ```suggestion
* Deprecate `KOKKOS_NONTEMPORAL_PREFETCH_{LOAD,STORE}` macros [\#8146](https://github.com/kokkos/kokkos/pull/8146)
```

### Bug Fixes

> * Get the concurrency from provided execution space instances [\#7870](https://github.com/kokkos/kokkos/pull/7870)

Reviewer 2: ```suggestion
* Get the concurrency from provided execution space instances in CUDA/HIP `parallel_for(TeamPolicy)` and CUDA/HIP/SYCL `UniqueToken` with global scope [\#7870](https://github.com/kokkos/kokkos/pull/7870)
```

* Fix simd math functions not compiling when ARCH_NATIVE=ON [\#7912](https://github.com/kokkos/kokkos/pull/7912)

> * HIP and SYCL: fixed a bug where a `MDRangePolicy` of rank 4 or more is incorrectly iterated, leading to some iterations being evaluated more than once for large loops [\#7880](https://github.com/kokkos/kokkos/pull/7880)

Reviewer 2: Move that one first, that is probably the most important bug

Author: Rearranged.

* Fix memory leak in internals of the Serial default execution space [\#8042](https://github.com/kokkos/kokkos/pull/8042)

> * Fix memory leak in the initialization functions [\#8074](https://github.com/kokkos/kokkos/pull/8074)

Reviewer 2: ```suggestion
* Fix memory leaks when initializing with `--kokkos-tools-help` [\#8074](https://github.com/kokkos/kokkos/pull/8074)
```
I think technically that was the only issue that we had in 4.6
@tretre91 is that right

* HIP: fix a bug a where `ConstantMemory` launch mechanism sporadically fails due to `hipEventSynchronize` error [\#8094](https://github.com/kokkos/kokkos/pull/8094)
* Fix fallback simd masked reductions using incorrect identity elements [\#8115](https://github.com/kokkos/kokkos/pull/8115)

> * HPX fix to constrain hpx_thread_buffer size used with TeamPolicy setup [\#8147](https://github.com/kokkos/kokkos/pull/8147)

Reviewer 2: What does that mean not in terms of internals but as far as users are concerned?
```suggestion
* HPX: Corrected an internal TeamPolicy constraint that could cause issues with higher thread counts [\#8147](https://github.com/kokkos/kokkos/pull/8147)
```

* Fix subview view constructor to disallow non-strided view types for strided subviews [\#8210](https://github.com/kokkos/kokkos/pull/8210)
* Work around a compiler bug with GCC 12.{2,3,4} causing segfaults with the View shared allocation tracker [\#8223](https://github.com/kokkos/kokkos/pull/8223)

> * Fix compiling with compilers that provide an mdspan implementation [\#8234](https://github.com/kokkos/kokkos/pull/8234)

Reviewer 2: Would that actually trigger an issue with 4.6?  Did you check?

* Fix launch of intermediate size functors in graph on HIP [\#8188](https://github.com/kokkos/kokkos/pull/8188)

> * Add missing fences for async Random init with unified memory [\#8105](https://github.com/kokkos/kokkos/pull/8105)

Reviewer 2: ```suggestion
* Add missing fences for async random pool init with unified memory [\#8105](https://github.com/kokkos/kokkos/pull/8105)
```
(I don't feel too strongly about that one)

> * OpenMPTarget / OpenACC: fix out-of-bounds access in `Random_UniqueIndex` under certain circumstances [\#8077](https://github.com/kokkos/kokkos/pull/8077)

Reviewer 2: `Random_UniqueIndex` is in `Impl::`
Can you suggest something less cryptic?

Author: We could elaborate the description a bit more to be something like:
```
OpenMPTarget / OpenACC: Fix an out-of-bounds access error in `Kokkos Random` pool initialization when thread count exceeds the maximum concurrency of an execution space.
```

# CHANGELOG

## 4.7.00

[Full Changelog](https://github.com/kokkos/kokkos/compare/4.6.02...4.7.00)

### Features:
* Introduce mdspan-based View implementation [\#7427](https://github.com/kokkos/kokkos/pull/7427)
  * This change should be transparent for most users, but it is a complete reimplementation of `View`

### Backend and Architecture Enhancements:

#### CUDA:
* Add support for AMPERE87 architecture (Jetson Orin Nano) [\#8092](https://github.com/kokkos/kokkos/pull/8092)
* Support RDC with Clang 17+ and use new offload driver [\#7831](https://github.com/kokkos/kokkos/pull/7831)
* Directly use fp16 mathematical function when available instead of casting back and forth to fp32 [\#7698](https://github.com/kokkos/kokkos/pull/7698)
* Disable `mdspan`-based view for CUDA 12.9 due to compiler incompatibility[\#8169](https://github.com/kokkos/kokkos/pull/8169)

#### SYCL:
* Add support for Intel DG2 GPUs such as the Arc Alchemist GPUs [\#7922](https://github.com/kokkos/kokkos/pull/7922)

> * Allow requesting out-of-order queues [\#8091](https://github.com/kokkos/kokkos/pull/8091)

Reviewer: ```suggestion
```

* Remove `nodiscard` attributes from Kokkos simd interface [\#8084](https://github.com/kokkos/kokkos/pull/8084)
* Add support for graph capture (`Cuda`, `HIP` and `SYCL`) [\#7552](https://github.com/kokkos/kokkos/pull/7552)
* Enforce unit launch bound for graph `then` node [\#8192](https://github.com/kokkos/kokkos/pull/8192)
* Add `then_host` to graph [\#8075](https://github.com/kokkos/kokkos/pull/8075)
* Allow building `Kokkos::Experimental::Graph` object directly [\#8108](https://github.com/kokkos/kokkos/pull/8108)

### Build System Changes
* Set GCC minimum version for C++20 to 10.4 [\#8130](https://github.com/kokkos/kokkos/pull/8130)
* Disallow relocatable device code for shared library build [\#8196](https://github.com/kokkos/kokkos/pull/8196)
* Check if compiler/linker can consume the flags set [\#7891](https://github.com/kokkos/kokkos/pull/7891)
* Support more nvcc arguments with nvcc_wrapper [\#7930](https://github.com/kokkos/kokkos/pull/7930)
* Add NVIDIA Blackwell architecture support to the generated makefiles [\#8055](https://github.com/kokkos/kokkos/pull/8055)

> * Add support for building C++20 modules [\#8132](https://github.com/kokkos/kokkos/pull/8132)

Reviewer: ```suggestion
* Add initial C++20 modules configuration support [\#8132](https://github.com/kokkos/kokkos/pull/8132)
```

* Fix simd math functions not compiling when ARCH_NATIVE=ON [\#7912](https://github.com/kokkos/kokkos/pull/7912)
* Fix memory leak in internals of the Serial default execution space [\#8042](https://github.com/kokkos/kokkos/pull/8042)
* Fix memory leaks when initializing with `--kokkos-tools-help` [\#8074](https://github.com/kokkos/kokkos/pull/8074)
* HIP: fix a bug a where `ConstantMemory` launch mechanism sporadically fails due to `hipEventSynchronize` error [\#8094](https://github.com/kokkos/kokkos/pull/8094)
* Fix fallback simd masked reductions using incorrect identity elements [\#8115](https://github.com/kokkos/kokkos/pull/8115)
* HPX: Corrected an internal TeamPolicy constraint that could cause issues with higher thread counts [\#8147](https://github.com/kokkos/kokkos/pull/8147)
* Fix subview view constructor to disallow non-strided view types for strided subviews [\#8210](https://github.com/kokkos/kokkos/pull/8210)
* Work around a compiler bug with GCC 12.{2,3,4} causing segfaults with the View shared allocation tracker [\#8223](https://github.com/kokkos/kokkos/pull/8223)

> * Fix compiling with compilers that provide an mdspan implementation [\#8234](https://github.com/kokkos/kokkos/pull/8234)

Reviewer: ```suggestion
```

This doesn't really work anyway.

Reviewer 4: The point is rather that we can compile with a compiler that provides an mdspan implementation even our configuration implied that we didn't mean to use it.

### Features:

### Backend and Architecture Enhancements:

#### CUDA:
* Support RDC with Clang 17+ and use new offload driver [\#7831](https://github.com/kokkos/kokkos/pull/7831)
* Add support for AMPERE87 architecture (Jetson Orin Nano) [\#8092](https://github.com/kokkos/kokkos/pull/8092)
* Directly use fp16 mathematical function when available instead of casting back and forth to fp32 [\#7698](https://github.com/kokkos/kokkos/pull/7698)

#### HIP:

#### SYCL:
- Allow requesting out-of-order queues [\#8091](https://github.com/kokkos/kokkos/pull/8091)
- Allow using non-trivially-copyable comparators with oneDPL [\#7538](https://github.com/kokkos/kokkos/pull/7538)
- Directly use fp16 mathematical function when available instead of casting back and forth to fp32 [\#7698](https://github.com/kokkos/kokkos/pull/7698)

#### OpenMPTarget:
- Remove support for non-llvm compilers [\7973](https://github.com/kokkos/kokkos/pull/7973)

#### OpenACC:

#### HPX:

#### Threads:

#### OpenMP:

#### Serial:

### General Enhancements
* Add support for Zen 5 AMD microarchitecture [\#7877](https://github.com/kokkos/kokkos/pull/7877)

> * Add support for NVIDIA's Blackwell GPU architecture [\#7888](https://github.com/kokkos/kokkos/pull/7888)

Reviewer 4: Wasn't this added in 4.6.1?
```suggestion
```

Author: Yes, looks like this was already added into 4.6.1. Removed.

* Add support for SiFive Unmatched Rev B (RISC-V)  [\#7921](https://github.com/kokkos/kokkos/pull/7921) 

> * Add support for Intel DG2 GPUs such as the Arc Alchemist GPUs [\#7922](https://github.com/kokkos/kokkos/pull/7922)

Reviewer 4: Could also be under SYCL if we mention #8092 under CUDA. We should be consistent.

Author: Moved `#7922` under SYCL section.

* Add support for graph capture (`Cuda`, `HIP` and `SYCL`) [\#7552](https://github.com/kokkos/kokkos/pull/7552)
* Improve performance of `atomic_op_fetch()` [\#8014](https://github.com/kokkos/kokkos/pull/8014)
* Pass label by reference in all `Kokkos::tools` APIs [\#8065](https://github.com/kokkos/kokkos/pull/8065)
* Compatibility with C+20 modules: Avoid static variables and functions in header files [\#8071](https://github.com/kokkos/kokkos/pull/8071)
* Exit early at initialize with `--kokkos-help` [\#8079](https://github.com/kokkos/kokkos/pull/8079)
* Add Kokkos::SIMD::SVE support for 128-bit and 256-bit SVE [#7807](https://github.com/kokkos/kokkos/pull/7807)
* Add constructors for Random_XorShift*_Pool with execution space argument [\#8039](https://github.com/kokkos/kokkos/pull/8039)
* Implement `nextafter` for fp16 [\#8118](https://github.com/kokkos/kokkos/pull/8118)

> * Enable structured binding return for `parition_space(ExecSpace, Args...)` [\#8114](https://github.com/kokkos/kokkos/pull/8114)

Reviewer 4: Also an incompatibility (no changes needed but we should clraify in the release briefing).

* Enforce unit launch bound for graph `then` node [\#8192](https://github.com/kokkos/kokkos/pull/8192)
* Add `constexpr` specifier to `operator==` and `operator!=` for `Kokkos::complex` [\#8254](https://github.com/kokkos/kokkos/pull/8254)

> * Add `then_host` to graph [\#8075](https://github.com/kokkos/kokkos/pull/8075)

Reviewer 4: It probbaly makes sense to list all Graph changes next to each other.

Author: Clumped all Graph PRs together.

* Support more nvcc arguments with nvcc_wrapper [\#7930](https://github.com/kokkos/kokkos/pull/7930)
* Add NVIDIA Blackwell architecture support to the generated makefiles
[\#8055](https://github.com/kokkos/kokkos/pull/8055)
* Add support for building C++20 modules [\#8132](https://github.com/kokkos/kokkos/pull/8132)
* Set GCC minimum version for cxx 20 [\#8130](https://github.com/kokkos/kokkos/pull/8130)

### Incompatibilities (i.e. breaking changes)
* Raise IntelLLVM minimum for SYCL to 2024.2.1 [\#7918](https://github.com/kokkos/kokkos/pull/7918)
* Disallow `Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF` [\#8021](https://github.com/kokkos/kokkos/pull/8021)

### Deprecations
* Deprecate KOKKOS_MEMORY_ALIGNMENT, KOKKOS_MEMORY_ALIGNMENT_THRESHOLD [\#7873](https://github.com/kokkos/kokkos/pull/7873)
*  Deprecate Kokkos::MemoryManaged [\#8066](https://github.com/kokkos/kokkos/pull/8066)
*  Deprecate `KOKKOS_NONTEMPORAL_PREFETCH_[LOAD|STORE]` [\#8146](https://github.com/kokkos/kokkos/pull/8146)

### Bug Fixes
* Get the concurrency from provided execution space instances [\#7870](https://github.com/kokkos/kokkos/pull/7870)

> * Make sure that AUTO has external linkage for C++20 modules support [\#7898](https://github.com/kokkos/kokkos/pull/7898)

Reviewer 4: Part of 4.6.1.
```suggestion
```

Author: Removed.

* Fix simd math functions not compiling when ARCH_NATIVE=ON [\#7912](https://github.com/kokkos/kokkos/pull/7912)
* HIP and SYCL: fixed a bug where a `MDRangePolicy` of rank 4 or more is incorrectly iterated, leading to some iterations being evaluated more than once for large loops [\#7880](https://github.com/kokkos/kokkos/pull/7880)

> * Building with generated makefiles: add missing HIP XNACK source file [\#8030](https://github.com/kokkos/kokkos/pull/8030)

Reviewer 4: ```suggestion
```

Author: Removed.
